### PR TITLE
Add operation details

### PR DESCRIPTION
**Pull request by mbj**
### Acceptance

* [x] Verify `stack test --fast`

### CODE REVIEW

File: src/OpenApi/Paths.hs [hunk 1]
```diff
@@ -1,24 +1,37 @@
+{-# OPTIONS_GHC -Wwarn #-}
```
Comment (mbj) [inline src/OpenApi/Paths.hs]: leftover to be removed.

File: src/OpenApi/Paths.hs [hunk 2]
```diff
@@ -30,11 +41,85 @@ newtype OperationId = OperationId Text
   deriving newtype (JSON.FromJSON, ToText)
   deriving stock   Show
 
+data Responses = Responses
+  { default' :: Maybe Response
+  , patterns :: Map ResponseStatusPattern Response
+  }
+  deriving stock Show
+
+instance JSON.FromJSON Responses where
```
Comment (mbj) [inline src/OpenApi/Paths.hs:50]: This instance needs a test, blocking.

File: test/Main.hs [hunk 3]
```diff
@@ -19,17 +19,20 @@ parseTemplate
...
+      , ("/foo",         Paths.Template [static "foo"])
+      , ("/foo/bar",     Paths.Template [static "foo", static "bar"])
+      , ("/foo/{bar}",   Paths.Template [static "foo", dynamic "bar"])
+      , ("/foo_bar",     Paths.Template [static "foo_bar"])
+      , ("/{foo}",       Paths.Template [dynamic "foo"])
+      , ("/{foo}/bar",   Paths.Template [dynamic "foo", static "bar"])
+      , ("/{foo}/{bar}", Paths.Template [dynamic "foo", dynamic "bar"])
       ]
 
+    static = Paths.Static
```
Comment (snusnu) [inline test/Main.hs]: @mbj The next line should likely be aligned on the `=`.

**Comment by snusnu** [comment]
@mbj Not a blocker, but the commit message for

* `Change dynamic segment name to be more precise`

confused me slightly initially. I was looking for a more precise *name*, when in fact, afaict, you actually changed the dynamic segment name *type*  to be more precise. Wdyt about reflecting this in the commit message, ie

* `Change dynamic segment name type to be more precise`

File: src/OpenApi/HTTP.hs [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+import Numeric.Natural (Natural)
+import OpenApi.Prelude
+
+import qualified Network.HTTP.Types.Status as HTTP
+
+-- Convert safely from natural to known http status.
+--
+-- Both `toEnum` and upstream `mkStatus` allow to create odd
+-- statuses, like negative or unregistred ones.
```
Comment (snusnu) [inline src/OpenApi/HTTP.hs]: `s/unregistred/unregistered`. I am going to fix this myself.

File: src/OpenApi/Paths.hs [hunk 5]
```diff
@@ -1,24 +1,35 @@
...
+import qualified Data.HashMap.Strict       as HashMap
+import qualified Data.Map.Strict           as Map
+import qualified GHC.Enum                  as GHC
+import qualified Network.HTTP.Types.Status as HTTP
 
 data Operation = Operation
   { description :: Maybe (Description Operation)
   , operationId :: OperationId
   , parameters  :: Maybe [Parameter]
+  , deprecated  :: Maybe Bool
```
Comment (snusnu) [inline src/OpenApi/Paths.hs]: This list of record fields should be sorted. I'm going to fix this myself.

File: src/OpenApi/Paths.hs [hunk 6]
```diff
@@ -30,6 +41,79 @@ newtype OperationId = OperationId Text
...
+        :: Text
+        -> JSON.Value
+        -> JSON.Parser (ResponseStatusPattern, Response)
+      parsePair key value =
+        (,) <$> parseResponseStatusPattern key <*> JSON.parseJSON value
+
+      parseResponseStatusPattern :: Text -> JSON.Parser ResponseStatusPattern
+      parseResponseStatusPattern input =
+        either
+          (const . fail $ "invalid status code pattern: " <> show input)
```
Comment (snusnu) [inline src/OpenApi/Paths.hs]: I think that for consistency, the initial capitalization should match that on LOC 75 (ie. read `Invalid ...`). I'm going to fix this myself.

File: src/OpenApi/HTTP.hs [hunk 7]
```diff
@@ -0,0 +1,62 @@
+module OpenApi.HTTP where
```
Comment (snusnu) [inline src/OpenApi/HTTP.hs:1]: @mbj Afaict, `module`s with an empty export list export all the contained functions. I see that this `module` only defines a single function that will be exported, but I wonder if, for maintenance reasons, we should start nailing the exports down precisely? The `OpenApi.HTTP` module is maybe not the most convincing module for this flag, but I chose to flag here as this is the first green line occurrence in this commit where I noticed this, the `OpenApi.Paths` module is also "affected".

At this point I'm mainly curious if we already have an established rule wrt such default exports? I tend to think that relying on exporting everything by default may lead to undesired exports as the codebase grows?

Comment (mbj) [reply src/OpenApi/HTTP.hs:1]: @snusnu I do not consider the import / export list mature here at all. Potentially I'll not expose all but `OpenApi` root and do not export internal modules at all.

For now I only added export lists to things where:

* I do not need everything exported (this is AFAIK a case, and re-specifying everything where the default is everything: is just redundant).
* I'am sure about the future public interface.

This is not "mature" code. It needs to grow into our usual standards.

**Comment by mbj** [comment]
@dkubb, @snusnu this is ready for review.

File: test/Main.hs [hunk 8]
```diff
@@ -60,5 +62,56 @@ parseTemplate
     parseJSON :: Text -> Either String Paths.Template
     parseJSON = JSON.parseEither JSON.parseJSON . JSON.String
 
+parseResponses :: TestTree
+  = testGroup "OpenAPI.Path.Responses parsing"
+  $ accepted <> rejected
+  where
+    accepted :: [TestTree]
+    accepted = mkAccepted <$>
+      [ (JSON.object [], Paths.Responses { default' = empty, patterns = Map.empty })
```
Comment (dkubb) [inline test/Main.hs]: The `[]` here and elsewhere in this test can be changed to `empty`.

File: test/Main.hs [hunk 9]
```diff
@@ -60,5 +62,56 @@ parseTemplate
...
+        response = Paths.Response
+          { description = empty
+          , headers     = empty
+          , content     = Map.empty
+          }
+
+    responsesJSON :: JSON.Value
+    responsesJSON =
+      JSON.object
+        [ ("200", JSON.object [("content", JSON.object [])])]
```
Comment (dkubb) [inline test/Main.hs]: The `]` at the end of the line is missing a leading space; which will make it consistent with the other code in this commit/file.

File: src/OpenApi/Paths.hs [hunk 10]
```diff
@@ -30,6 +41,79 @@ newtype OperationId = OperationId Text
...
+        (,) <$> parseResponseStatusPattern key <*> JSON.parseJSON value
+
+      parseResponseStatusPattern :: Text -> JSON.Parser ResponseStatusPattern
+      parseResponseStatusPattern input =
+        either
+          (const . fail $ "Invalid status code pattern: " <> show input)
+          pure $ Text.parseOnly statusCodeParser input
+
+      statusCodeParser =
+        maybe (fail "Invalid status code") (pure . ResponseStatusExact)
```
Comment (dkubb) [inline src/OpenApi/Paths.hs:75]: Not a blocker, but given `either` is on it's own line above, I would probably put `maybe` on it's own line too.

Comment (dkubb) [reply src/OpenApi/Paths.hs:75]: Actually, I see now how the whole expression fits on one line, and the line below is the input feeding into the `maybe`. Please disregard.

**Review by dkubb — APPROVED** [review]

**Pull request merged**